# INTRODUCTION

In [111]:
# importing necessary libraries
import pandas as pd
import matplotlib as mp
import seaborn as sb



In [231]:
# Accessing the dataframe
retail_df = pd.read_excel('online_retail.xlsx', sheet_name="original")

# Dataset Description
**InvoiceNo:** Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 

**StockCode:** Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.

**Description:** Product (item) name. Nominal.

**Quantity:** The quantities of each product (item) per transaction. Numeric.	

**InvoiceDate:** Invice Date and time. Numeric, the day and time when each transaction was generated.

**UnitPrice:** Unit price. Numeric, Product price per unit in sterling.

**CustomerID:** Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.

**Country:** Country name. Nominal, the name of the country where each customer resides. 

# INVESTIGATING THE DATAFRAME

In [232]:
# Investigating the columns in the dataframe
retail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [233]:
## Checking for duplicates
sum(retail_df.duplicated())

5268

In [234]:
retail_df.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
229622    557066     22469                HEART OF WICKER SMALL        12   
387333    570284     22834          HAND WARMER BABUSHKA DESIGN         1   
165845    550829     23133     LARGE IVORY HEART WALL ORGANISER         1   
37763     539477    84247G             DECOUPAGE,GREETING CARD,        12   
528545    580729     23282      FOLDING BUTTERFLY MIRROR IVORY          2   
475643    576909     23158          SET OF 5 LUCKY CAT MAGNETS         12   
336507    566345    85199L  LARGE HANGING IVORY & RED WOOD BIRD        48   
503999    578941     84991          60 TEATIME FAIRY CAKE CASES         2   
338931    566571     22568                FELTCRAFT CUSHION OWL         4   
509483    579307     22639        SET OF 4 NAPKIN CHARMS HEARTS         6   

               InvoiceDate  UnitPrice  CustomerID         Country  
229622 2011-06-16 15:10:00       1.65     17463.0  United Kingdom  
387333 2011-10-10 11:30:00       2.10     14419.0  United Kingdom  
165845 2011-04-21 10:19:00       8.25     17841.0  United Kingdom  
37763  2010-12-19 14:58:00       0.42     18245.0  United Kingdom  
528545 2011-12-05 17:24:00       2.46         NaN  United Kingdom  
475643 2011-11-17 09:49:00       2.08     12471.0         Germany  
336507 2011-09-12 12:27:00       0.65     17685.0  United Kingdom  
503999 2011-11-27 13:26:00       0.55     16024.0  United Kingdom  
338931 2011-09-13 12:49:00       3.75     17037.0  United Kingdom  
509483 2011-11-29 11:34:00       2.55     17655.0  United Kingdom

In [235]:
# checking for the number of records without CustomerID
retail_df_noCustomerID = len(retail_df[retail_df['CustomerID'].isnull()])
print(retail_df_noCustomerID)

135080


## Issues in the Dataframe
1. It can be seen that there are 5268 duplicate records 
2. It is observed that some orders doesn't have CustomerID
3. it can be seen that CustomerID column is a datatype float64

# CLEANING UP

## Issue 1: DUPLICATED RECORDS

## Operation: To remove duplicate records

### Code:

In [236]:
# removing duplicated records
retail_df.drop_duplicates(inplace=True)

### Test:

In [237]:
sum(retail_df.duplicated())

0

### Comment:

Duplicated records have now being removed

## Issue 2: SOME ORDERS DOESNT HAVE CUSTOMER ID

## Operation: To remove orders with no customer ID

Since the overhaul purpose of this project is to create customer segmentation using the dataset only Invoice which contains CustomerID is needed

### Code:

In [238]:
## Creating a dataframe of only records with CustomerID
df_with_CustomerID = retail_df[retail_df['CustomerID'].notnull()]

In [239]:
df_with_CustomerID.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.6+ MB


### Test:

In [240]:
sum(df_with_CustomerID['CustomerID'].isnull())

0

### Comment

As shown above records without customerID have been removed

## Issue 3: CUSTOMERID COLUMN IS IN FLOAT DATATYPE

There is need to change the customerid column datatype from float to string since there is no possibility of carrying out arithmetic operations on this column

### Code

In [241]:
new_df = df_with_CustomerID.copy()

In [250]:
new_df['CustomerID'] = df_with_CustomerID['CustomerID'].astype(int).astype(object)


### Test:

In [244]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  object        
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 27.6+ MB


### Comment:

CustomerID datatype have now being changed to string

## SAVING THE CLEAN DATAFRAME

In [253]:
new_df.to_csv("cleaned_dataset.csv", index=False)